[TechDebt]: Add `PreCheck` to CloudSearch domain based tests

### Description

In an effort to shift towards the OpenSearch service, AWS has recently introduced a change which prevents accounts with no active CloudSearch domains from creating new ones. The relevant context from the support case:

> We introduced a change on July 25, wherein new accounts that don't already have one or more CloudSearch domains, will have to be explicitly allowlisted for new domain creation. If you would like to create a CloudSearch domain on an AWS account other than those that you are using today, we can allowlist it for new domain creation. This change does not affect existing CloudSearch domains, and you can continue to create new domains on existing accounts in which you already have CloudSearch domains running. 

Acceptance tests for this service which rely on a domain should be updated with a PreCheck to skip on `NotAuthorized` exceptions. 

```
=== RUN   TestAccCloudSearchDomain_basic
=== PAUSE TestAccCloudSearchDomain_basic
=== CONT  TestAccCloudSearchDomain_basic
    domain_test.go:28: Step 1/2 error: Error running apply: exit status 1
        Error: creating CloudSearch Domain (tf-acc-test-owp77qhzuy982v49): operation error CloudSearch: CreateDomain, https response error StatusCode: 401, RequestID: d75b043a-25ab-47d6-b62f-6925fa1097ae, api error NotAuthorized: New domain creation not supported on this account. Please reach out to AWS Support for assistance.
          with aws_cloudsearch_domain.test,
          on terraform_plugin_test.tf line 12, in resource "aws_cloudsearch_domain" "test":
          12: resource "aws_cloudsearch_domain" "test" {
--- FAIL: TestAccCloudSearchDomain_basic (31.01s)
FAIL
```

The `ListDomainNames` API can be used as the PreCheck.

```console
% aws cloudsearch list-domain-names

An error occurred (NotAuthorized) when calling the ListDomainNames operation: New domain creation not supported on this account. Please reach out to AWS Support for assistance.
```

### References

Relates https://github.com/hashicorp/terraform-provider-aws/pull/38509#issuecomment-2256587266

### Would you like to implement a fix?

None

--- a/internal/acctest/acctest.go
+++ b/internal/acctest/acctest.go
@@ -24,6 +24,7 @@ import (
 	accounttypes "github.com/aws/aws-sdk-go-v2/service/account/types"
 	"github.com/aws/aws-sdk-go-v2/service/acmpca"
 	acmpcatypes "github.com/aws/aws-sdk-go-v2/service/acmpca/types"
+	"github.com/aws/aws-sdk-go-v2/service/cloudsearch"
 	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
 	"github.com/aws/aws-sdk-go-v2/service/costexplorer"
 	"github.com/aws/aws-sdk-go-v2/service/directoryservice"
@@ -1039,6 +1040,22 @@ func PreCheckCECostCategoryPayerAccount(ctx context.Context, t *testing.T) {
 
 	if err != nil {
 		t.Fatalf("listing Cost Explorer Cost Categories: %s", err)
+	}
+}
+
+func PreCheckCloudSearchAccountAllowListed(ctx context.Context, t *testing.T) {
+	t.Helper()
+
+	conn := Provider.Meta().(*conns.AWSClient).CloudSearchClient(ctx)
+
+	_, err := conn.ListDomainNames(ctx, &cloudsearch.ListDomainNamesInput{})
+
+	if errs.MessageContains(err, "NotAuthorized", "New domain creation not supported on this account") {
+		t.Skip("skipping tests; this AWS account does not support new CloudSearch domain creation")
+	}
+
+	if err != nil {
+		t.Fatalf("listing CloudSearch Domain Names: %s", err)
 	}
 }
 

--- a/internal/service/cloudsearch/domain_service_access_policy_test.go
+++ b/internal/service/cloudsearch/domain_service_access_policy_test.go
@@ -24,7 +24,11 @@ func TestAccCloudSearchDomainServiceAccessPolicy_basic(t *testing.T) {
 	rName := acctest.ResourcePrefix + "-" + sdkacctest.RandString(28-(len(acctest.ResourcePrefix)+1))
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainServiceAccessPolicyDestroy(ctx),
@@ -51,7 +55,11 @@ func TestAccCloudSearchDomainServiceAccessPolicy_update(t *testing.T) {
 	rName := acctest.ResourcePrefix + "-" + sdkacctest.RandString(28-(len(acctest.ResourcePrefix)+1))
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainServiceAccessPolicyDestroy(ctx),

--- a/internal/service/cloudsearch/domain_test.go
+++ b/internal/service/cloudsearch/domain_test.go
@@ -26,7 +26,11 @@ func TestAccCloudSearchDomain_basic(t *testing.T) {
 	rName := testAccDomainName()
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainDestroy(ctx),
@@ -65,7 +69,11 @@ func TestAccCloudSearchDomain_disappears(t *testing.T) {
 	rName := testAccDomainName()
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainDestroy(ctx),
@@ -89,7 +97,11 @@ func TestAccCloudSearchDomain_indexFields(t *testing.T) {
 	rName := testAccDomainName()
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainDestroy(ctx),
@@ -161,7 +173,11 @@ func TestAccCloudSearchDomain_sourceFields(t *testing.T) {
 	rName := testAccDomainName()
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainDestroy(ctx),
@@ -236,7 +252,11 @@ func TestAccCloudSearchDomain_update(t *testing.T) {
 	rName := testAccDomainName()
 
 	resource.ParallelTest(t, resource.TestCase{
-		PreCheck:                 func() { acctest.PreCheck(ctx, t); acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID) },
+		PreCheck: func() {
+			acctest.PreCheck(ctx, t)
+			acctest.PreCheckPartitionHasService(t, names.CloudSearchEndpointID)
+			acctest.PreCheckCloudSearchAccountAllowListed(ctx, t)
+		},
 		ErrorCheck:               acctest.ErrorCheck(t, names.CloudSearchServiceID),
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
 		CheckDestroy:             testAccCheckDomainDestroy(ctx),

